### ISSUE TITLE
Consumer doesn't get messages anymore

### ISSUE THREAD

**Issue by calohmn** [eclipse/hono#232]
In our Hono setup we have a consumer connected to the dispatch router. There is no log output indicating a lost connection.
At one point, the consumer didn't get messages anymore.

Logs from the Hono server/messaging component show that the connection to QPid was lost and re-established in-between:
```
05:03:24.641 [vert.x-eventloop-thread-1] WARN o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - lost connection to downstream container [Hono.Example.Router], closing upstream receivers ...
05:03:24.641 [vert.x-eventloop-thread-1] WARN o.e.h.e.i.ForwardingEventDownstreamAdapter - lost connection to downstream container [Hono.Example.Router], closing upstream receivers ...
05:03:25.142 [vert.x-eventloop-thread-1] INFO o.e.h.t.i.ForwardingTelemetryDownstreamAdapter - attempting to re-connect to downstream container
05:03:25.142 [vert.x-eventloop-thread-1] INFO o.e.h.s.DownstreamClientConfigProperties - using certificates from file [/run/secrets/trusted-certs.pem] as trust anchor
05:03:25.142 [vert.x-eventloop-thread-1] INFO o.e.h.s.DownstreamClientConfigProperties - using key & certificate from PKCS12 key store [/run/secrets/honoKeyStore.p12] for identity
05:03:25.142 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qdrouter.hono:5673]
05:03:25.149 [vert.x-eventloop-thread-1] INFO o.e.h.e.i.ForwardingEventDownstreamAdapter - attempting to re-connect to downstream container
05:03:25.149 [vert.x-eventloop-thread-1] INFO o.e.h.s.DownstreamClientConfigProperties - using certificates from file [/run/secrets/trusted-certs.pem] as trust anchor
05:03:25.149 [vert.x-eventloop-thread-1] INFO o.e.h.s.DownstreamClientConfigProperties - using key & certificate from PKCS12 key store [/run/secrets/honoKeyStore.p12] for identity
05:03:25.149 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qdrouter.hono:5673]
05:03:25.272 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connected to AMQP 1.0 container [amqps://qdrouter.hono:5673], opening connection ...
05:03:25.274 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connection to container [Hono.Example.Router] at [amqps://qdrouter.hono:5673] open
05:03:25.277 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connected to AMQP 1.0 container [amqps://qdrouter.hono:5673], opening connection ...
05:03:25.278 [vert.x-eventloop-thread-1] INFO o.e.h.c.ConnectionFactoryImpl - connection to container [Hono.Example.Router] at [amqps://qdrouter.hono:5673] open
```
More log output here:
[Hono_messaging.txt](https://github.com/eclipse/hono/files/1097407/Hono_messaging.txt)

A 2nd consumer didn't show this behaviour, it still got messages.

**Comment by sophokles73** [comment]
is the 2nd consumer receiving from the same source address as the one that doesn't work any more after the disconnect?

**Comment by calohmn** [comment]
We have further analyzed the scenario and found that the problem occurs because the disconnect handler given to `HonoClient.connect()` isn't invoked again after one successful re-connect.

The issue can be reproduced by running the Example Receiver and stopping/restarting the Dispatch Router connection twice.
After the second QPID restart, the consumer doesn't get telemetry messages anymore.

Log output:
```
12:13:19.402 [main] INFO  org.eclipse.hono.example.Application - The following profiles are active: receiver,ssl
12:13:19.585 [main] INFO  o.s.c.a.AnnotationConfigApplicationContext - Refreshing org.springframework.context.annotation.AnnotationConfigApplicationContext@1a78822f: startup date [Tue Jun 27 12:13:19 CEST 2017]; root of context hierarchy
12:13:20.079 [main] INFO  org.eclipse.hono.example.Application - Starting example client in role receiver,ssl
12:13:20.371 [main] INFO  o.e.h.config.ClientConfigProperties - using certificates from file [target/certs/trusted-certs.pem] as trust anchor
12:13:20.372 [main] INFO  o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qpid:15671]
12:13:20.962 [main] INFO  o.s.j.e.a.AnnotationMBeanExporter - Registering beans for JMX exposure on startup
12:13:20.971 [main] INFO  org.eclipse.hono.example.Application - Started Application in 2.005 seconds (JVM running for 5.853)
12:13:21.870 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connected to AMQP 1.0 container [amqps://qpid:15671], opening connection ...
12:13:21.908 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connection to container [Hono.Example.Router] at [amqps://qpid:15671] open
12:13:21.952 [vert.x-eventloop-thread-0] INFO  o.e.hono.example.ExampleReceiver - Receiver [tenant: DEFAULT_TENANT, type: telemetry] created successfully, hit ctrl-c to exit

12:13:22.673 [vert.x-eventloop-thread-0] INFO  o.e.hono.example.ExampleReceiver - received telemetry message ...

[QPID stopped and restarted for the first time]

12:13:53.292 [vert.x-eventloop-thread-0] INFO  o.e.hono.client.impl.HonoClientImpl - lost connection to server [qpid:15671]
12:13:53.494 [vert.x-eventloop-thread-0] INFO  o.e.hono.example.ExampleReceiver - attempting to re-connect to Hono server
12:13:53.495 [vert.x-eventloop-thread-0] INFO  o.e.h.config.ClientConfigProperties - using certificates from file [target/certs/trusted-certs.pem] as trust anchor
12:13:53.495 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qpid:15671]
12:14:03.525 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - can't connect to AMQP 1.0 container [amqps://qpid:15671]: handshake timed out
12:14:04.027 [vert.x-eventloop-thread-0] INFO  o.e.hono.client.impl.HonoClientImpl - attempting to re-connect to server [qpid:15671]
12:14:04.027 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qpid:15671]
12:14:14.068 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - can't connect to AMQP 1.0 container [amqps://qpid:15671]: handshake timed out
12:14:14.571 [vert.x-eventloop-thread-0] INFO  o.e.hono.client.impl.HonoClientImpl - attempting to re-connect to server [qpid:15671]
12:14:14.572 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qpid:15671]
12:14:14.760 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connected to AMQP 1.0 container [amqps://qpid:15671], opening connection ...
12:14:14.782 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connection to container [Hono.Example.Router] at [amqps://qpid:15671] open

12:14:16.681 [vert.x-eventloop-thread-0] INFO  o.e.hono.example.ExampleReceiver - received telemetry message

[QPID stopped and restarted for the second time]

12:14:24.238 [vert.x-eventloop-thread-0] INFO  o.e.hono.client.impl.HonoClientImpl - lost connection to server [qpid:15671]
12:14:24.741 [vert.x-eventloop-thread-0] INFO  o.e.hono.client.impl.HonoClientImpl - attempting to re-connect to server [qpid:15671]
12:14:24.741 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qpid:15671]
12:14:35.059 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - can't connect to AMQP 1.0 container [amqps://qpid:15671]: handshake timed out
12:14:35.560 [vert.x-eventloop-thread-0] INFO  o.e.hono.client.impl.HonoClientImpl - attempting to re-connect to server [qpid:15671]
12:14:35.563 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qpid:15671]
12:14:45.604 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - can't connect to AMQP 1.0 container [amqps://qpid:15671]: handshake timed out
12:14:46.105 [vert.x-eventloop-thread-0] INFO  o.e.hono.client.impl.HonoClientImpl - attempting to re-connect to server [qpid:15671]
12:14:46.105 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qpid:15671]
12:14:56.138 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - can't connect to AMQP 1.0 container [amqps://qpid:15671]: handshake timed out
12:14:56.640 [vert.x-eventloop-thread-0] INFO  o.e.hono.client.impl.HonoClientImpl - attempting to re-connect to server [qpid:15671]
12:14:56.640 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qpid:15671]
12:15:06.667 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - can't connect to AMQP 1.0 container [amqps://qpid:15671]: handshake timed out
12:15:07.169 [vert.x-eventloop-thread-0] INFO  o.e.hono.client.impl.HonoClientImpl - attempting to re-connect to server [qpid:15671]
12:15:07.169 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connecting to AMQP 1.0 container [amqps://qpid:15671]
12:15:07.642 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connected to AMQP 1.0 container [amqps://qpid:15671], opening connection ...
12:15:07.661 [vert.x-eventloop-thread-0] INFO  o.e.h.c.ConnectionFactoryImpl - connection to container [Hono.Example.Router] at [amqps://qpid:15671] open

[no further telemetry messages received]

```

**Comment by sophokles73** [comment]
The `HonoClient.connect(ProtonClientOptions, Handler, Handler)` variant provides a way to supply your own *disconnect handler*, which is invoked when the connection unexpectedly fails. You can provide a handler which basically invokes `HonoClient.connect()` again using the original *connection handler*.

The current API does not make any claims for the original *connection handler* being invoked again after a re-connect has occurred. I am not sure if this what a client would want to happen in the first place. That's what the variant of `HonoClient.connect()` mentioned above is intended to be used for ...

**Comment by calohmn** [comment]
> The HonoClient.connect(ProtonClientOptions, Handler, Handler) variant provides a way to supply your own disconnect handler, which is invoked when the connection unexpectedly fails. You can provide a handler which basically invokes HonoClient.connect() again using the original connection handler.

Yes, that is what we did and what is also done in `org.eclipse.hono.example.ExampleReceiver`.

> The current API does not make any claims for the original connection handler being invoked again after a re-connect has occurred. I am not sure if this what a client would want to happen in the first place.

What we want and what I guess is the most common scenario, is a consumer application that still gets telemetry messages and events after the connection to the Dispatch Router got lost several times. This means being able to ensure that `honoClient.createTelemetryConsumer()` is invoked after a re-connect. 

For the HonoClient to support this, I think it should either invoke the _connection handler_ on a re-connect, or the `HonoClient` API should support an extra _re-connect handler_.

**Comment by sophokles73** [comment]
I am not sure if I am following.
> Yes, that is what we did and what is also done in org.eclipse.hono.example.ExampleReceiver.

Are you saying that this *does not* work?

**Comment by sophokles73** [comment]
should be fixed in 43468221cbef4a3fbbc48849791c8b32b1b65777
